src/parser/jobs/whm/modules/Thinair.tsx

```diff
@@ -0,0 +1,143 @@
+import {t} from '@lingui/macro'
+import {Plural, Trans} from '@lingui/react'
+import {ActionLink} from 'components/ui/DbLink'
+import Rotation from 'components/ui/Rotation'
+import {getDataBy} from 'data'
+import ACTIONS from 'data/ACTIONS'
+import STATUSES from 'data/STATUSES'
+import {Event, Events} from 'event'
+import {Analyser} from 'parser/core/Analyser'
+import {filter} from 'parser/core/filter'
+import React, {Fragment, ReactNode} from 'react'
+import {Accordion} from 'semantic-ui-react'
+
+type ThinAirRecord = {
+	start: number,
+	end: number,
+	casts: number[],
+	mpsaved: number,
+}
+
+export default class Thinair extends Analyser {
+	static override handle = 'thinair'
+	static override title = t('whm.thinair.title')`Thin Air`
+
+	private active = false
+	private history: ThinAirRecord[] = []
+	private currentRecord: ThinAirRecord = {start: -1, end: -1, casts: [], mpsaved: 0}
+
+	override initialise() {
+		const thinairFilter = filter<Event>()
+			.source(this.parser.actor.id)
+			.status(STATUSES.THIN_AIR.id)
+
+		this.addEventHook(thinairFilter.type('statusApply'), this.onThinAirCast)
+		this.addEventHook(thinairFilter.type('statusRemove'), this.onThinAirRemove)
+		this.addEventHook(filter<Event>().source(this.parser.actor.id).type('action'), this.onCast)
+		this.addEventHook('complete', this.onComplete)
+	}
+
+	private onThinAirCast(ev: Events['statusApply']) {
+		this.startThinAir(ev.timestamp)
+	}
+
+	private onThinAirRemove(ev: Events['statusRemove']) {
+		this.stopAndSave(ev.timestamp)
+	}
+
+	private onCast(ev: Events['action']) {
+		const actionid = ev.action
+
+		if (actionid === ACTIONS.THIN_AIR.id) {
+			this.startThinAir(ev.timestamp)
+		}
+
+		// Don't track autos
+		const action = getDataBy(ACTIONS, 'id', actionid)
+		if (!this.active || !action || action.autoAttack) {
+			return
+		}
+
+		if (action.onGcd) {
+			const mpcost = action.mpCost === undefined ? 0 : action.mpCost
+			this.currentRecord.mpsaved += mpcost
+		}
```

Comment:
While all current WHM spells with MP cost happen to be on the GCD, that isn't necessarily a guarantee going forward as other classes have resource consuming oGCDs.  If you change this check to ensure there's an mpCost on the action, you'll simplify the remaining logic accordingly.

```suggestion
		if (action.mpCost != null) {
			this.currentRecord.mpsaved += action.mpCost
		}
```